### ISSUE TITLE
Dual universal sentence encoder?

### ISSUE THREAD

**Issue by ShellRox** [tensorflow/hub#201]
I'm trying to create a semantic similarity search algorithm based on cosine similarity. 

The simplest thing I can do, is to use universal sentence encoder to encode all the textual information tokenized in sentences to 512 dimensional embedding vector space, and perform sequential query on it. But it is highly inefficient. 

Hence I thought I would find some algorithm that can be used to quickly search through high dimensional spaces (such as locality sensitive hashing or random partition forests dividing space into convex polyhedrons). But then on [Semantic Experiences I found this](https://research.google.com/semanticexperiences/for-developers.html):

> The Universal Sentence Encoder model is very similar to what we're using in Talk to Books and Semantris, although those applications are using a dual-encoder approach that maximizes for response relevance, while the Universal Sentence Encoder is a single encoder that returns an embedding for the input, instead of a score on an input pair.

Is there any implementation of this? Can I modify Universal Sentence Encoder easily to become a dual-encoder? Or do I have to create my own Deep Averaging Neural Network (or Transformer) at fundamental level that does this?

Thank you!

**Comment by vbardiovskyg** [comment]
Hi,

I don't think the text suggests that having a dual encoder would make it somehow faster to perform a semantic similarity search. If you what you are searching for and what you search from are from the same distribution (sentences), then single encoder should be enough.

How to build a search algorithm is a different question. In tensorflow you could use:
```
distances = tf.matmul(my_sentence_embedding, other_embeddings)
topk_distances, topk_indices = tf.nn.top_k(distances)
```

Obviously it is not very efficient for online search, so the other solutions that you proposed (locality sensitive hashing, etc...) make probably much more sense.

**Comment by ShellRox** [comment]
> Hi,
> 
> I don't think the text suggests that having a dual encoder would make it somehow faster to perform a semantic similarity search. If you what you are searching for and what you search from are from the same distribution (sentences), then single encoder should be enough.
> 
> How to build a search algorithm is a different question. In tensorflow you could use:
> 
> ```
> distances = tf.matmul(my_sentence_embedding, other_embeddings)
> topk_distances, topk_indices = tf.nn.top_k(distances)
> ```
> Obviously it is not very efficient for online search, so the other solutions that you proposed (locality sensitive hashing, etc...) make probably much more sense.

I've actually researched about locality sensitive hashing and even found slightly more accurate and faster methods (random binary partition forests dividing vector space into convex polyhedrons). 

But I thought dual encoder could make slight changes in speed and accuracy, but it seems like I'm wrong.

**Comment by MajorMajorMajorMajor** [comment]
> > The text indirectly states that "dual-encoder [provides] a score on an input pair". Single encoder gives just an embedding. So there is probably an accuracy win ("maximizes for response relevance"), but it is not saying that this would be a performance optimization. Even if you get score for a pair, you still have to try all the pairs (or be somehow smart about it).
> 
> Now it is completely understandable! I'll keep working on the searching function!
> 
> Thank you very much for the help!

Any luck finding an algorithm to avoid matching every pair?
